packages/keyring-controller/src/KeyringController.test.ts

```diff
@@ -3080,6 +3081,71 @@ describe('KeyringController', () => {
             },
           );
         });
+
+        it('should update the vault if the keyring is being updated', async () => {
+          await withController(
+            { keyringBuilders: [keyringBuilderFactory(MockMutableKeyring)] },
+            async ({ controller, messenger }) => {
+              const selector = { type: MockMutableKeyring.type };
+
+              const mockStateChange = jest.fn();
+              messenger.subscribe(
+                'KeyringController:stateChange',
+                mockStateChange,
+              );
+
+              await controller.withKeyring(
+                selector,
+                async ({ keyring }) => {
+                  (keyring as MockMutableKeyring).update(); // Update the keyring state.
+                },
+                {
+                  createIfMissing: true,
+                },
+              );
+
+              expect(mockStateChange).toHaveBeenCalled();
+              expect(controller.state.keyrings).toHaveLength(2);
+            },
+          );
+        });
```

Comment:
nit: We can probably accomplish the same result without adding a `MockMutableKeyring` and by looking at what happens to the vault specifically through the encryptor:

```suggestion
        it('should update the vault if the keyring is being updated', async () => {
          stubKeyringClassWithAccount(MockKeyring, '0x1234');
          await withController(
            { keyringBuilders: [keyringBuilderFactory(MockKeyring)] },
            async ({ controller, encryptor }) => {
              await controller.addNewKeyring(MockKeyring.type);
              const encryptSpy = jest.spyOn(encryptor, 'encrypt');
              jest.spyOn(MockKeyring.prototype, 'serialize').mockResolvedValue({
                foo: 'bar',
              });

              await controller.withKeyring(
                { type: MockKeyring.type },
                async () => {
                  // No-op
                },
              );

              expect(encryptSpy).not.toHaveBeenCalled();
            },
          );
        });
```

Reply:
I did something similar to your suggestion but kept my use of the `messenger` since I think it's more "explicit" for my use-case: https://github.com/MetaMask/core/pull/5732/commits/676415f047cc21f5df81534aa4070e62e5775ccb